### PR TITLE
Implemented racks/create, racks/delete

### PR DESCRIPTION
**Pull request by juliasaveliff**
Following implementation decisions were made:
- On create:
   - If some/all racks within specified range already exist, request fails and no changes are made (does NOT specify which racks exist)
- On delete:
   - If some/all racks within specified range do not exist, request fails and no changes are made (does NOT specify which racks do not exist) 
   - If some/all racks within specified range have instances in them, request fails and no changes are made (specifies which racks have instances in them) 

Subject to change according to response at: https://piazza.com/class/k4u27qnccr45oo?cid=50

### CODE REVIEW

**Comment by benauriemma** [comment]
You said "If some/all racks within specified range do not exist, request fails and no changes are made (does NOT specify which racks do not exist)" but I thought we were returning a string with info about that?

**Comment by benauriemma** [comment]
Did/can you document those decisions somewhere for posterity?

File: rackcity/views/rack_views.py [hunk 1]
```diff
@@ -76,7 +76,31 @@ def rack_create(request):
     """
     Create racks within specified range.
     """
-    return JsonResponse({}, status=HTTPStatus.INTERNAL_SERVER_ERROR)
+    range_serializer = RackRangeSerializer(data=request.data)
+    if not range_serializer.is_valid():
+        return JsonResponse(
+            {"failure_message": "Invalid keys or values in request."},
```
Comment (benauriemma) [inline rackcity/views/rack_views.py]: I think you can do `{"failure_message": str(range_serializer.errors)}` here can't you?

File: rackcity/views/rack_views.py [hunk 2]
```diff
@@ -85,4 +109,48 @@ def rack_delete(request):
     """
     Delete racks within specified range.
     """
-    return JsonResponse({}, status=HTTPStatus.INTERNAL_SERVER_ERROR)
+    range_serializer = RackRangeSerializer(data=request.data)
+    if not range_serializer.is_valid():
+        return JsonResponse(
+            {"failure_message": "Invalid keys or values in request."},
```
Comment (benauriemma) [inline rackcity/views/rack_views.py]: Same as above

File: rackcity/views/rack_views.py [hunk 3]
```diff
@@ -85,4 +109,48 @@ def rack_delete(request):
...
+    if len(unempty_racks) > 0:
+        failure_message = "No racks were deleted because the following " + \
+            "racks contain instances: " + ", ".join(unempty_racks)
+        return JsonResponse(
+            {"failure_message": failure_message},
+            status=HTTPStatus.BAD_REQUEST,
+        )
+    else:
+        try:
+            racks.delete()
```
Comment (benauriemma) [inline rackcity/views/rack_views.py]: You can call this on a list of models? Wild

File: rackcity/views/rack_views.py [hunk 4]
```diff
@@ -76,7 +76,31 @@ def rack_create(request):
...
+            status=HTTPStatus.BAD_REQUEST,
+        )
+    racks = Rack.objects.filter(
+        rack_num__range=range_serializer.get_number_range(),  # inclusive range
+        row_letter__range=range_serializer.get_row_range(),
+    )
+    if racks.count() > 0:
+        failure_message = "Racks cannot be created because racks within " + \
+            "the range already exist."
+        return JsonResponse(
```
Comment (benauriemma) [inline rackcity/views/rack_views.py]: Seems like you're following the fail early pattern here which we don't really do elsewhere. I don't dislike it but worth tracking and considering standardizing one way or the other at some point down the line

**Comment by benauriemma** [comment]
How crazy is it that we did vooga with like 4 times as many people and were definitely was worse engineers at the time but simply had no code review protocol? Like you could really just merge whatever you wanted straight to master

File: rackcity/views/rack_views.py (hunk 1 above)
Comment (juliasaveliff) [reply rackcity/views/rack_views.py]: so true. fixing this

File: rackcity/views/rack_views.py (hunk 3 above)
Comment (juliasaveliff) [reply rackcity/views/rack_views.py]: p what? p wild

File: rackcity/views/rack_views.py (hunk 4 above)
Comment (juliasaveliff) [reply rackcity/views/rack_views.py]: tracking engaged

**Comment by juliasaveliff** [comment]
@benauriemma aDdEd cHanGeS cAn YOu pLeASe rEviEW aNd apProvE

**Comment by juliasaveliff** [comment]
Milestone: I just accidentally closed a PR

File: rackcity/views/rack_views.py (hunk 2 above)
Comment (benauriemma) [reply rackcity/views/rack_views.py]: Bippity bump

Comment (benauriemma) [reply rackcity/views/rack_views.py]: Wait I take it back I think you fixed this

**Pull request merged**